lib/filterx/filterx-object.h

```diff
@@ -0,0 +1,174 @@
+/*
+ * Copyright (c) 2023 Balazs Scheidler <balazs.scheidler@axoflow.com>
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 2.1 of the License, or (at your option) any later version.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
+ *
+ * As an additional exemption you are allowed to compile & link against the
+ * OpenSSL libraries as published by the OpenSSL project. See the file
+ * COPYING for details.
+ *
+ */
+#ifndef FILTERX_OBJECT_H_INCLUDED
+#define FILTERX_OBJECT_H_INCLUDED
+
+#include "logmsg/logmsg.h"
+#include <json-c/json.h>
+
+typedef struct _FilterXType FilterXType;
+typedef struct _FilterXObject FilterXObject;
+
+struct _FilterXType
+{
+  FilterXType *super_type;
+  const gchar *name;
+  gboolean mutable;
+  union
+  {
+    struct
+    {
+      FilterXObject *(*unmarshal)(FilterXObject *self);
+      gboolean (*marshal)(FilterXObject *self, GString *repr, LogMessageValueType *t);
+      FilterXObject *(*clone)(FilterXObject *self);
+      gboolean (*map_to_json)(FilterXObject *self, struct json_object **object);
+      gboolean (*truthy)(FilterXObject *self);
+      FilterXObject *(*getattr)(FilterXObject *self, const gchar *attr_name);
+      gboolean (*setattr)(FilterXObject *self, const gchar *attr_name, FilterXObject *new_value);
+      void (*free_fn)(FilterXObject *self);
+    };
+    gpointer __methods[8];
+  };
+};
+
+void filterx_type_init(FilterXType *type);
+
+#define FILTERX_TYPE_NAME(_name) filterx_type_ ## _name
+#define FILTERX_DECLARE_TYPE(_name) \
+    extern FilterXType FILTERX_TYPE_NAME(_name)
+#define FILTERX_DEFINE_TYPE(_name, _super_type, ...) \
+    FilterXType FILTERX_TYPE_NAME(_name) = 	\
+    {						\
+      .super_type = &_super_type,		\
+      .name = # _name,				\
+      __VA_ARGS__				\
+    }
+
+
+FILTERX_DECLARE_TYPE(object);
+
+struct _FilterXObject
+{
+  /* NOTE: this packs into 16 bytes in total (64 bit), let's try to keep
+   * this small, potentially using bitfields.  A simple boolean is 32 bytes
+   * in total at the moment (factoring in GenericNumber which is used to
+   * represent it).  Maybe we could get rid off the GenericNumber wrapper
+   * which would potentially decrease the struct to 16-24 bytes. */
+
+  gint ref_cnt;
```

Comment:
I think this should be atomic, we had hard-to-reproduce bugs when we forgot about non-atomic refcounts and started extending lifetimes across thread boundaries.

Reply:
filterx objects should not cross thread boundaries. That would be a lot of overhead of making all of the refcounts atomic. A lot more of this churn than refcounts of anything else in syslog-ng.

Reply:
I made a stab at this by adding an assert into filterx_object_unref(), please review that. I am leaving this note open.